test/reshape.jl

```diff
@@ -654,6 +654,74 @@ end
 end
 
+@testset "unstack with fillvalue" begin
+    df = DataFrame(factory=["Fac1", "Fac1", "Fac2", "Fac2"],
+                   variable=["Var1", "Var2", "Var1", "Var2"],
+                   value=[1, 2, 3, 4])
+    dfu1 = DataFrame(factory=["Fac1", "Fac2"],
+                     Var1=allowmissing([1, 3]),
+                     Var2=allowmissing([2, 4]))
+    dfu = unstack(df, :variable, :value)
+    @test dfu ≅ dfu1
+    @test eltype(dfu.Var1) === Union{Missing, Int}
+    @test eltype(dfu.Var2) === Union{Missing, Int}
+
+    for (sentinel, coleltype) in zip([1, 1., "1", nothing], [Int, Float64, Any, Union{Int, Nothing}])
+        dfu = unstack(df, :variable, :value, fillvalue=sentinel)
+        @test dfu ≅ dfu1
+        @test eltype(dfu.Var1) === coleltype
+        @test eltype(dfu.Var2) === coleltype
+    end
+
+    df = DataFrame(factory=["Fac1", "Fac1", "Fac2"],
+                   variable=["Var1", "Var2", "Var1"],
+                   value=[1, 2, 3])
+    for (sentinel, coleltype) in zip([1, 1.0, "1", nothing], [Int, Float64, Any, Union{Int, Nothing}])
+        dfu = unstack(df, :variable, :value, fillvalue=sentinel)
+        @test dfu.Var1 == [1, 3]
+        @test eltype(dfu.Var1) === coleltype
+        @test dfu.Var2 == [2, sentinel]
+        @test eltype(dfu.Var2) === coleltype
+    end
+
+    df = DataFrame(factory=["Fac1", "Fac1", "Fac2"],
+                   variable=["Var1", "Var2", "Var1"],
+                   value=categorical([1, 2, 3], ordered=true))
+    # categorical is dropped here
+    for (sentinel, coleltype) in zip([0, 0.0, "", nothing], [Int, Float64, Any, Union{Int, Nothing}])
+        dfu = unstack(df, :variable, :value, fillvalue=sentinel)
+        @test dfu.Var1 == [1, 3]
+        @test typeof(dfu.Var1) === Vector{coleltype}
+        @test dfu.Var2 == [2, sentinel]
+        @test typeof(dfu.Var2) === Vector{coleltype}
+    end
+    # categorical is kept here
+    for (sentinel, coleltype) in zip([missing, CategoricalValue(1, df.value), ], [Union{Int, Missing}, Int])
+        dfu = unstack(df, :variable, :value, fillvalue=sentinel)
+        @test dfu.Var1 == [1, 3]
+        @test typeof(dfu.Var1) <: CategoricalVector{coleltype}
+        @test dfu.Var2 ≅ [2, sentinel]
+        @test typeof(dfu.Var2) <: CategoricalVector{coleltype}
+        @test levels(dfu.Var1) == levels(dfu.Var2) == levels(df.value)
+    end
+
+    df = DataFrame(factory=["Fac1", "Fac1", "Fac2"],
+                   variable=["Var1", "Var2", "Var1"],
+                   value=categorical([1, 2, 3]))
+    dfu = unstack(df, :variable, :value, fillvalue=CategoricalValue(0, categorical([0])))
+    @test dfu.Var1 == [1, 3]
+    @test typeof(dfu.Var1) <: CategoricalVector{Int}
+    @test dfu.Var2 ≅ [2, 0]
+    @test typeof(dfu.Var2) <: CategoricalVector{Int}
+    @test levels(dfu.Var1) == levels(dfu.Var2) == 0:3
+    dfu = unstack(df, :variable, :value, fillvalue=CategoricalValue("0", categorical(["0"])))
```

Comment:
@nalimilan - this test in particular is something we should think of. My judgement is that it is OK to produce such an union.